### PR TITLE
KAFKA-15265: Add Remote Log Manager quota manager

### PR DESCRIPTION
**Pull request by abhijeetk88**
In this PR, I have added the implementation of the quota manager that will be used to throttle copy and fetch requests from the remote storage. Reference [KIP-956](https://cwiki.apache.org/confluence/display/KAFKA/KIP-956+Tiered+Storage+Quotas)

Added unit-tests for the Quota Manager implementation

In follow-up PRs, the integration of the copy/fetch quota managers with copy and fetch paths will be added. Also the quota configs will be made dynamically configurable.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (HenryCaiHaiying) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]: LOGGER.warn ?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (funky-eyes) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:1]: I have a question, in fact, what this PR does is to provide a standard configuration and generate corresponding rate limiters and related monitoring indicators, right? Then it needs to be used in the corresponding RemoteStorageManager, correct?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]: Makes sense, will change this.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:1]: Yes, the integration of the quota manager will come in the follow-up PRs. I have mentioned it in the description of the PR.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java:19]: do we need to add `toString()`?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:85]: metricName contains the quota-type. Do we need to log the quota-type similar to other quota-manager implementations?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]:
```suggestion
    KafkaMetric metric = allMetrics.get(quotaMetricName);
    if (metric != null) {
        Map<String, String> metricTags = quotaMetricName.tags();
        LOGGER.info("Sensor for quota-id {} already exists. Setting quota to {} in MetricConfig", metricTags, newQuota);
        metric.config(getQuotaMetricConfig(newQuota));
    }
```
why are we not accessing the map directly similar to other implementations? (ie) why `metricTags` is excluded?

File: core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java
Comment (kamalcph) [inline core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java]: consider swapping: 

assertEquals(actual, expected) to assertEquals(expected, actual)

File: core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java
Comment (kamalcph) [inline core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java]: consider swapping:

assertEquals(actual, expected) to assertEquals(expected, actual)

**Review by kamalcph — COMMENTED** [review]
Overall LGTM, left minor comments to address. Thanks for the patch!

File: core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java
Comment (abhijeetk88) [reply core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java]: Yes, thanks for pointing out.

File: core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java
Comment (abhijeetk88) [reply core/src/test/java/kafka/log/remote/quota/RLMQuotaManagerTest.java]: Yes, thanks.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java:19]: Will add.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]: Ack will change this. Also, metricTags is empty. Will fix the log statement as well.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java:19]: Added

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:85]: quota type is already being captured in the sensor instance name. So we can skip logging it separately.

**Review by kamalcph — APPROVED** [review]
LGTM

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: `This is a global limit for all the partitions that are being copied from remote storage to local storage.` <-- is it right? Copied from local storage to remote storage?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (showuon) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]: I'd like to know why we set WARN logs here. It looks to me if we want to update quota dynamically, it is expected the metric is already existed, right? If so, I don't think this is unexpected. So maybe INFO or DEBUG level, WDYT?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (showuon) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:39]: I can see this class is quite similar with `ReplicationQuotaManager`. Could you please create a JIRA issue to see if we could refactor them to avoid duplication in the future? Thanks.

**Review by showuon — COMMENTED** [review]
Overall LGTM. Left some comments.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (junrao) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java]: For consistency, we don't typically use getters. So this can just be quotaBytesPerSecond. Ditto below.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (junrao) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:93]: Why do we turn the quota checking off during `record()`? In other implementations like `ClientQuotaManager.recordAndGetThrottleTimeMs()`, we call `record()` by turning on quota checking and get back the amount of time to throttle.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (junrao) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java:199]: Could we add the javadoc for the new param?

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (junrao) [inline storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:171]: Why do we choose the default fetch window different from copy?

**Review by junrao — COMMENTED** [review]
@abhijeetk88 : Thanks for the PR. Made a pass of non-testing files. Left a few comments.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (abhijeetk88) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: My bad. Thanks for catching this.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java]: On second thought, INFO should be the right level, because quota update is a significant change in the application state and will decide how fast copies/fetches from remote storage can happen. Also, quota updates are infrequent, hence it will not cause excessive logging.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:39]: Created https://issues.apache.org/jira/browse/KAFKA-16706

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java]: Sure, will change these.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:199]: Sure, will add.

**Comment by showuon** [comment]
@abhijeetk88 , do we have any update on this PR?

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java:93]: In KIP-956, we do not utilize the throttle time provided by the quota manager to regulate fetches and copies. For fetch operations, we initially verify quota availability before initiating the retrieval of remote data. If the quota is unavailable, our priority is to serve partitions requiring local data, rather than throttling the client. Therefore, we focus on fulfilling data requests for other partitions in the queue, eliminating the need for throttle time in fetch operations.

Similarly, when a RLM Task attempts to copy a segment, it first checks if the write quota is available. If the quota is not available, the thread waits until the quota becomes available. As a result, we do not require throttle time for copies either.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (abhijeetk88) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:171]: For fetches, the default window size was chosen to match the default window size used for other quotas, such as ClientQuota and ReplicationQuota.

Using an 11-second (10 whole + 1 current) window size for copies, similar to other quotas, does seem to be a better option. Consider this:

The broker-level quota for copying may be set to 250 MBps. The RLM task records the log segment size with the quota manager when uploading a log segment. The typical log segment size is 500 MB, meaning only one log segment can be uploaded every 2 seconds without breaching the quota. If uploads occur faster, the quota will be exceeded. Therefore, as long as the window size is greater than 2 seconds, either a 10-second or 60-second (whole) window should work.

However, a shorter window (10 seconds) has advantages. It tracks data uploads more precisely and prevents large spikes in data upload more effectively. For example:

With a 10-second window:

Buckets: b1, b2, ..., b10
In the 10th second, 5 segments can be uploaded without breaching the average quota (5 * 500 MB / 10 seconds = 250 MBps), though the spike will be 2.5 GB in that second.
With a 60-second window:

Buckets: b1, b2, ..., b60
In the 60th second, 30 segments can be uploaded without breaching the average quota (30 * 500 MB / 60 seconds = 250 MBps), but the spike will be 15 GB in that second.
Given the need to avoid quota breaches, a 10-second window is preferable to a 60-second window.

Let me know if it makes sense. I can change the default copy window to be the same as the default fetch window.

**Comment by abhijeetk88** [comment]
@showuon @junrao I have addressed your comments. Please take a look.

**Review by showuon — APPROVED** [review]
LGTM! Thanks.

**Review by satishd — APPROVED** [review]
Thanks @abhijeetk88 for addressing the review comments. LGTM.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java:33]: nit: 

place constructor before the getter methods

**Review by kamalcph — COMMENTED** [review]
Do we plan to change the default copy quota window num samples from 61 to 11?

**Comment by showuon** [comment]
@abhijeetk88 , there is a merge conflict. Please help resolve it. Thanks.

File: core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/quota/RLMQuotaManagerConfig.java:33]: Done

**Comment by abhijeetk88** [comment]
> @abhijeetk88 , there is a merge conflict. Please help resolve it. Thanks.

done


> Do we plan to change the default copy quota window num samples from 61 to 11?

Waiting for a confirmation from Jun.

**Comment by showuon** [comment]
@junrao , since this PR blocks other follow-up PRs and v3.8.0 release date is approaching, I'd like to merge it tomorrow if you don't have any other comments. Thanks.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (junrao) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:171]: If there is no good reach, perhaps it's better to use the same default window number for copy too.

**Review by junrao — COMMENTED** [review]
@abhijeetk88 : Sorry for the delay. Just a minor followup comment. Otherwise, the PR LGTM.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java
Comment (abhijeetk88) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:171]: Done

**Review by junrao — APPROVED** [review]
@abhijeetk88 : Thanks for the updated PR. LGTM

**Comment by jolshan** [comment]
Can we look at testCopyQuotaManagerConfig() – kafka.log.remote.RemoteLogManagerTest? It seems like it is failing pretty consistently. https://issues.apache.org/jira/browse/KAFKA-16866

**Comment by chia7712** [comment]
@jolshan I file https://github.com/apache/kafka/pull/16146 to fix it.

**Comment by junrao** [comment]
@jolshan : Thanks for pointing this out. Sorry that I didn't look at the test results carefully before merging.

**Comment by abhijeetk88** [comment]
Thanks @chia7712 @jolshan . Apologies for the miss.

**Comment by gaobosince1987** [comment]
@abhijeetk88 trying the quota feature in test cluster now from trunk, primarily to save CPU when multiple consumer pull at the same time. Is there a document related with how to set quota for remote store? Latest trunk build does not seem to throttle remote fetch for me. Let me know if there is extra setup needed for this feature to work.

I have a 6 broker test cluster, 1 test topic with local.retention.ms=1 so all read goes to remote. However consumer are still able to fetch 300MB per broker from remote. With the configuration, I would expect close to zero remote fetch.
 [root@kafka-test-02-zookeeper-0 kafka]#./bin/kafka-configs.sh --bootstrap-server kafka-test-02-kafka-bootstrap:9092 --entity-type brokers --entity-default --alter --add-config 'remote.log.manager.fetch.max.bytes.per.second=1,remote.log.manager.copy.max.bytes.per.second=1428800'
Completed updating default config for brokers in the cluster.

**Pull request merged**
